### PR TITLE
JDK11 distribution creation.

### PR DESCRIPTION
**Pull request by bobmcwhirter**
Much cleaner PR (will close the other) that can create a JDK11-based distribution without duplicating the `vm/` directory. Simply modified the mx scripts and launcher templates.

* Allow JDK11 in the version check.
* Only search for classpath entries not provided by the JDK.
* Adjust for presence or lack of jre/ within a distro.
* A module-centric launcher template.
* Guard module arguments with JDK11+ check.
* Defensiveness around the polyglot launcher class names.

### CODE REVIEW

**Comment by bobmcwhirter** [comment]
I don't understand that failure.

File: vm/mx.vm/mx_vm.py [hunk 1]
```diff
@@ -229,24 +234,39 @@ def _patch_darwin_jdk():
             vm_name = graalvm_vm_name(self, join(_jdk_dir, _src_jdk_base))
             if mx.get_os() == 'darwin':
                 # on macOS the <arch> directory is not used
-                _add(layout, "<jdk_base>/jre/lib/", "extracted-dependency:truffle:TRUFFLE_NFI_NATIVE/bin/<lib:trufflenfi>")
-                _add(layout, "<jdk_base>/jre/lib/server/vm.properties", "string:name=" + vm_name)
+                if _src_jdk_has_jre:
+                    _add(layout, "<jdk_base>/jre/lib/", "extracted-dependency:truffle:TRUFFLE_NFI_NATIVE/bin/<lib:trufflenfi>")
```
Comment (gilles-duboscq) [inline vm/mx.vm/mx_vm.py]: To avoid duplication we can introduce `<jre_base>` similar to `<jdk_base>`

**Comment by dougxc** [comment]
> I don't understand that failure.

I assume you mean:
```
testInstallCodeInvalidation(org.graalvm.compiler.hotspot.test.ObjectHashCodeInliningTest)
java.lang.AssertionError: failed HighTier graph check
	at org.graalvm.compiler.core.test.GraalCompilerTest$2.run(GraalCompilerTest.java:291)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.Phase.run(Phase.java:49)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:197)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:139)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.PhaseSuite.run(PhaseSuite.java:212)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:197)
	at jdk.internal.vm.compiler/org.graalvm.compiler.phases.BasePhase.apply(BasePhase.java:139)
	at jdk.internal.vm.compiler/org.graalvm.compiler.core.GraalCompiler.emitFrontEnd(GraalCompiler.java:220)
	at jdk.internal.vm.compiler/org.graalvm.compiler.core.GraalCompiler.compile(GraalCompiler.java:144)
	at org.graalvm.compiler.core.test.GraalCompilerTest.compile(GraalCompilerTest.java:1102)
	at org.graalvm.compiler.core.test.GraalCompilerTest.compile(GraalCompilerTest.java:1066)
	at org.graalvm.compiler.hotspot.test.ObjectHashCodeInliningTest.testInstallCodeInvalidation(ObjectHashCodeInliningTest.java:55)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
```
@mur47x111 this looks like a case of profile pollution caused by other tests running in the same VM.

File: vm/mx.vm/launcher_template_modules.sh [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+module_path="$(IFS=: ; echo "${absolute_module_path[*]}")"
+upgrade_module_path="$(IFS=: ; echo "${absolute_upgrade_module_path[*]}")"
+
+echo "ump: ${upgrade_module_path}"
+
+if [[ "${VERBOSE_GRAALVM_LAUNCHERS}" == "true" ]]; then
+    set -x
+fi
+
+echo exec "${location}/<jre_bin>/java" "${jvm_args[@]}" \
```
Comment (gilles-duboscq) [inline vm/mx.vm/launcher_template_modules.sh]: I guess this is left-over debug.
Note that it should also be possible debug this by setting the `VERBOSE_GRAALVM_LAUNCHERS` environment variable to `true`.

File: vm/mx.vm/mx_vm.py [hunk 3]
```diff
@@ -578,10 +631,28 @@ def native_image(self, build_args, output_file, allow_server=False, nonZeroIsFat
...
         if suffix:
             name = name[:-len(suffix)]
+
         native_image_command += [
             '-H:Path=' + output_directory or ".",
             '-H:Name=' + name,
         ]
+
+        if _get_jdk().version.parts[0] >= 11:
+            graal_sdk_jar = join(stage1.output, stage1.find_single_source_location('dependency:sdk:GRAAL_SDK'));
```
Comment (gilles-duboscq) [inline vm/mx.vm/mx_vm.py]: I'm surprised this is needed, i would expect the code guarded by `useJavaModules()` in `NativeImage.java` to already take care of whatever is necessary.
/cc @olpaw

**Comment by bobmcwhirter** [comment]
Thanks! I’ll make these changes.

File: vm/mx.vm/mx_vm.py (hunk 3 above)
Comment (olpaw) [reply vm/mx.vm/mx_vm.py]: I agree. This seems redundant/misguided. If any changes are necessary to build images with Java 11 they should be added to the already existing Java 11 support in native image. See:
https://github.com/oracle/graal/blob/af03ea82d35937f6d7bea8ec265ace6d88fffb94/substratevm/mx.substratevm/mx_substratevm.py#L138 (all definitions of `GRAAL_COMPILER_FLAGS_MAP['11']`)
https://github.com/oracle/graal/blob/af03ea82d35937f6d7bea8ec265ace6d88fffb94/substratevm/mx.substratevm/mx_substratevm.py#L1161
https://github.com/oracle/graal/blob/dde3d6fdd57256958b22621bbe74fd379507924c/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java#L117
https://github.com/oracle/graal/blob/dde3d6fdd57256958b22621bbe74fd379507924c/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java#L294
https://github.com/oracle/graal/blob/dde3d6fdd57256958b22621bbe74fd379507924c/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java#L223 (and search for all uses of `useJavaModules()`)

**Comment by bobmcwhirter** [comment]
I've integrated the above changes. 

I have not figured out how to supply `module-path` and `upgrade-module-path` so if you have a pointer on that, that'd be lovely.

File: vm/mx.vm/mx_vm.py [hunk 4]
```diff
@@ -105,16 +107,26 @@ def __init__(self, suite, name, deps, components, is_graalvm, exclLibs, platform
...
 
+        if _src_jdk_has_jre:
+            self.jre_base = '/'.join(self.jdk_base, 'jre')
+        else:
+            self.jre_base = self.jdk_base
+
+
         path_substitutions = mx_subst.SubstitutionEngine(mx_subst.path_substitutions)
         path_substitutions.register_no_arg('jdk_base', lambda: self.jdk_base)
+        path_substitutions.register_no_arg('jre_base', lambda: self.jdk_base)
```
Comment (gilles-duboscq) [inline vm/mx.vm/mx_vm.py]: i suppose this should be `self.jre_base`?

File: vm/mx.vm/mx_vm.py [hunk 5]
```diff
@@ -343,10 +356,11 @@ def _add_link(_dest, _target):
                     _add_link(_jdk_jre_bin, _link_dest)
                     _jre_bin_names.append(basename(_link_dest))
 
-            if 'jre' in _jdk_jre_bin:
-                # Add jdk to jre links
-                for _name in _jre_bin_names:
-                    _add_link('<jdk_base>/bin/', '<jdk_base>/jre/bin/' + _name)
+            if _src_jdk_has_jre:
+                if 'jre' in _jdk_jre_bin:
```
Comment (gilles-duboscq) [inline vm/mx.vm/mx_vm.py]: To avoid adding too much indentation depth, could you use `if ... and ...:`?

**Comment by gilles-duboscq** [comment]
> I have not figured out how to supply module-path and upgrade-module-path so if you have a pointer on that, that'd be lovely.

I assume you mean in `SvmSupport.native_image`? If yes, I'm suprized you would need to because `NativeImage` should already be doing it:
This should already be setting module-path and upgrade-module-path:
https://github.com/oracle/graal/blob/1159995eb78dac93c1ae6265c997595a51869cb4/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java#L664-L671
With the right jars:
https://github.com/oracle/graal/blob/1159995eb78dac93c1ae6265c997595a51869cb4/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java#L468-L469
https://github.com/oracle/graal/blob/1159995eb78dac93c1ae6265c997595a51869cb4/substratevm/src/com.oracle.svm.driver/src/com/oracle/svm/driver/NativeImage.java#L475

**Comment by bobmcwhirter** [comment]
Further adjusted.

Turns out the `NativeImage` was constructing paths that were:

* Looking in the wrong place (`lib/jvmci/graal-sdk.jar` instead of `lib/boot/` where the base layout places `graal-sdk.jar`)
* Not adding `graal-management.jar` to the module paths correctly.

I've removed the additional CLI args from `mx_vm.py`'s invocations.

File: vm/mx.vm/mx_vm.py (hunk 1 above)
Comment (bobmcwhirter) [reply vm/mx.vm/mx_vm.py]: done

File: vm/mx.vm/mx_vm.py (hunk 4 above)
Comment (bobmcwhirter) [reply vm/mx.vm/mx_vm.py]: yep, fixed.

**Comment by gilles-duboscq** [comment]
Great! Thank you for investigating this and going through the iterations.

This looks good to me, i'll integrate it.

**Comment by bobmcwhirter** [comment]
@gilles-duboscq Thanks!  

Once again CI is failing in some way that is obtuse to me.

If there's changes I need to do, help me unwind what Travis is telling me, please.

File: substratevm/mx.substratevm/mx_substratevm.py [hunk 6]
```diff
@@ -134,14 +138,22 @@ def add_opens_to_all_unnamed(packageName):
...
-    'org.graalvm.sdk/org.graalvm.nativeimage.impl']
+    'org.graalvm.sdk/org.graalvm.nativeimage.impl',
+    'org.graalvm.sdk/org.graalvm.polyglot',]
 GRAAL_COMPILER_FLAGS_MAP['11'].extend(add_opens_from_packages(graal_sdk_opens_packages))
 
+graal_truffle_opens_packages = [
+    'org.graalvm.truffle/com.oracle.truffle.polyglot',
+    'org.graalvm.truffle/com.oracle.truffle.api.impl',]
+GRAAL_COMPILER_FLAGS_MAP['11'].extend(add_opens_from_packages(graal_truffle_opens_packages))
+
```
Comment (gilles-duboscq) [inline substratevm/mx.substratevm/mx_substratevm.py]: One of the things Travis complains about is a trailing whitespace on this line (155)

File: vm/mx.vm/mx_vm.py [hunk 7]
```diff
@@ -105,16 +107,26 @@ def __init__(self, suite, name, deps, components, is_graalvm, exclLibs, platform
...
 
+        _src_jdk_has_jre = not (_src_jdk.version.parts[0] >= 11)
+
         if base_dir != '.':
             self.jdk_base = '/'.join([base_dir, _src_jdk_base]) if _src_jdk_base and _src_jdk_base != '.' else base_dir
         else:
             self.jdk_base = _src_jdk_base
 
+        if _src_jdk_has_jre:
+            self.jre_base = '/'.join(self.jdk_base, 'jre')
```
Comment (gilles-duboscq) [inline vm/mx.vm/mx_vm.py]: The second thing travis complains about is that `join` takes 1 iterable argument, not 2 arguments. i suppose you meant `'/'.join((self.jdk_base, 'jre'))

Comment (bobmcwhirter) [reply vm/mx.vm/mx_vm.py]: Ah, thanks. I'm still learning my way around Python.

**Comment by gilles-duboscq** [comment]
The last complain from travis is something going wrong when trying to build a native-image. Unfortunately, the actual error is swallowed.
The output indicates that the last command that failed was
```
mx --primary-suite-path /home/travis/build/oracle/graal/substratevm --java-home=/home/travis/oraclejdk11 native-image --no-fallback --tool:native-image -H:Path=/home/travis/build/oracle/graal/substratevm/svmbuild/native-image-root-11/linux-amd64/bin
```
Maybe that would be enough to reproduce locally? I think the actual error will be shown in verbose mode so maybe adding `--verbose` after `native-image ` in this command line will help.

**Comment by bobmcwhirter** [comment]
Thanks, I'll dig into that final failure.

**Comment by bobmcwhirter** [comment]
For posterity, the error seems to be

```
Error: Requirements for building native images are not fulfilled
com.oracle.svm.driver.NativeImage$NativeImageError: Requirements for building native images are not fulfilled
	at com.oracle.svm.driver.NativeImage.showError(NativeImage.java:1364)
	at com.oracle.svm.driver.NativeImage.build(NativeImage.java:1168)
	at com.oracle.svm.driver.NativeImage.performBuild(NativeImage.java:1144)
	at com.oracle.svm.driver.NativeImage.main(NativeImage.java:1103)
```

**Comment by gilles-duboscq** [comment]
I was hoping that with `--verbose` you would also get the root cause

**Comment by bobmcwhirter** [comment]
The root cause is that my changes look for `lib/boot/graal-sdk` even though in `svmbuild` they are in `lib/jvmci/`.

Trying to unwind this now, since they don't end up in `lib/jvmci/` under the `vm` distribution.

**Comment by bobmcwhirter** [comment]
Okay, graal-sdk now goes (back) into `lib/jvmci/` as expected, and `lib/boot/` is non-existent for JDK11 distros.  That final command passes locally now.

**Comment by bobmcwhirter** [comment]
I disclaim that LLVM failure. No idea about it.

**Comment by bobmcwhirter** [comment]
Added adjustments for native-image server, mostly in terms of adjusting `UNIXProcess` substitutions to also be applicable to JDK11's `ProcessImpl` name and structures.  Plus another couple of module adjustments for the server-side requirements.

**Comment by gilles-duboscq** [comment]
Hi Bob, i was going to integrate this but i see you're still making some changes.
I would prefer integrate this soon as a way to start prototyping on GraalVM/JDK11.

I saw you have formatting issues in the java code. The easiest way to get rid of those is to download Eclipse 4.5.2.1 (or something close enough), point the `ECLIPSE_EXE` environment variable to the `eclipse` executable, and run `mx eclipseformat`.
Since i'd like to integrate this soon, If you don't want to take care of formatting, i can also re-format in an other commit during integration.

**Comment by bobmcwhirter** [comment]
It’ll be a few hours before I can fix it. Please go ahead and merge and fix it up how you like. 

I was just adding changes since they were related but I can move to a new PR for subsequent ones. 

Carry on and thanks for the review and merge!

**Comment by bobmcwhirter** [comment]
Okay, just pushed an additional commit fixing formatting. `eclipseformat` wanted to touch a lot of unrelated files, so I have not committed those.

**Comment by gilles-duboscq** [comment]
> eclipseformat wanted to touch a lot of unrelated files, so I have not committed those.

That's usually the sign that you use a different version of eclipse.

FYI your old changes are already in the pipeline.

**Comment by bobmcwhirter** [comment]
Ah, okay. I saw the PR was still open (and needed to be rebased) so wasn't
sure.

-Bob

**Comment by gilles-duboscq** [comment]
Hi Bob, your commit prior to the rebase has been integrated.
As i offered, i did the formatting fix-ups in separate commits.

Maybe to clarify, this is the process on our end:
When a github PR is in good shape to be integrated,
- we pull the commits and create an internal PR
- we get a review from some more people
- we run gate tests
- If everything goes well, it gets integrated into master and pushed out. The integration is always a merge commit (no fast-forward).
- GitHub detects that the PR has been merged and marks it as such.

In this case since there was a rebase in between github didn't detect that the changes have been merge so i will close it manually.

**Comment by dmlloyd** [comment]
In general the rebase shouldn't be needed either: you could resolve the conflicts with merge commit(s).  In some ways this is better (at least in the case where the branch came from a contributor) because the branch being fixed up can have separate merge commits for each conflicting upstream commit, which makes it very clear how the contributed code needed to be modified for each upstream iteration.  This way the original commit IDs are preserved.

I would only use rebase in the case where the original commits are actually wrong, to avoid having bad commits in the tree (which spoils the efficacy of `git bisect`).  In that case if the fix is contributed by someone else and it's more than just whitespace or something, `Co-authored-by` is definitely appropriate.

Either way though, on GitHub it's best IMO to update the original PR with the fixed commits, either by the original contributor or by pushing directly to their branch (they will have had to have checked the "allow pushes from maintainers" box in this case).  This allows the contributor to continue testing with the final code instead of being stalled waiting for an opaque process.

**Comment by dougxc** [comment]
Thanks - I've updated our internal guidelines to refer to this useful advice.

**Pull request closed without merging**